src/Routing/PushState.purs

```diff
@@ -0,0 +1,241 @@
+import DOM.Node.Node (setNodeValue) as DOM
+import DOM.Node.Types (textToNode) as DOM
+import Data.Array as Array
+import Data.Foldable (class Foldable, for_, indexl, traverse_)
+import Data.Foreign (Foreign)
+import Data.Maybe (Maybe(..), fromMaybe, maybe)
+import Data.Newtype (class Newtype, unwrap)
+import Data.StrMap as StrMap
+import Data.String as String
+import Routing (match)
+import Routing.Match (Match)
+
+type PushStateEffects eff =
+  ( history :: HISTORY
+  , dom :: DOM
+  , ref :: REF
+  | eff
+  )
+
+newtype Basename = Basename String
+
+derive newtype instance eqBasename :: Eq Basename
+derive newtype instance ordBasename :: Ord Basename
+derive instance newtypeBasename :: Newtype Basename _
+
+newtype Path = Path String
+
+derive newtype instance eqPath :: Eq Path
+derive newtype instance ordPath :: Ord Path
+derive instance newtypePath :: Newtype Path _
+
+-- | A `PushStateInterface` is a localized instance for making location changes
+-- | and consuming the events. Since the DOM API does not provide a general
+-- | event type for push state (only pop), you must use the coupled effects
+-- | along with `listen` to receive events on all location changes.
+-- |
+-- | * `pushState` – pushes a new location state and path onto the history stack.
+-- | * `replaceState` – replaces the location state and path in the history stack.
+-- | * `locationState` – Dereferences the current history state
+-- | * `listen` – Subscribes to location changes (both push and pop). Returns
+-- |   an effect which removes the listener.
+type PushStateInterface eff =
+  { pushState :: Foreign -> Path -> Eff eff Unit
+  , replaceState :: Foreign -> Path -> Eff eff Unit
+  , locationState :: Eff eff LocationState
+  , listen :: (LocationState -> Eff eff Unit) -> Eff eff (Eff eff Unit)
+  }
+
+-- | Creates a new `PushStateInterface` given a `Basename`. The `Basename` is
+-- | used as an implicit prefix for the paths supplied to the instance, being
+-- | automatically added and stripped as necessary.
+makeInterface :: forall eff. Basename -> Eff (PushStateEffects eff) (PushStateInterface (PushStateEffects eff))
+makeInterface basename = do
+  argsRef <- newRef []
+  freshRef <- newRef 0
+  listenersRef <- newRef StrMap.empty
+
+  let
+    notify ev =
+      traverse_ (_ $ ev) =<< readRef listenersRef
+
+    listen k = do
+      fresh <- readRef freshRef
+      writeRef freshRef (fresh + 1)
+      modifyRef listenersRef $ StrMap.insert (show fresh) k
+      pure $ modifyRef listenersRef $ StrMap.delete (show fresh)
+
+    locationState = do
+      loc <- DOM.window >>= DOM.location
+      state <- DOM.window >>= DOM.history >>= DOM.state
+      pathname <- fromMaybe <*> String.stripPrefix (String.Pattern (unwrap basename)) <$> DOM.pathname loc
+      search <- DOM.search loc
+      hash <- DOM.hash loc
+      let path = pathname <> search <> hash
+      pure { state, pathname, search, hash, path }
+
+  schedule <- do
```

Comment:
Would be nice to inline-doc why we need to create a textNode and mutation-observe it, and what the "fresh" thing is.